services/spar/src/Spar/Scim.hs

```diff
@@ -194,37 +195,48 @@ validateScimUser'
 validateScimUser' Nothing _ =
     throwError $ Scim.serverError "SCIM users without SAML SSO are not supported"
 validateScimUser' (Just idp) user = do
-    let validateNameOrExtId :: Maybe Text -> m (Maybe Text)
-        validateNameOrExtId mtxt = forM mtxt $ \txt ->
-          case checkedEitherMsg @_ @1 @128 "displayName" txt of
-            Right rtxt -> pure $ fromRange rtxt
-            Left err -> throwError $ Scim.badRequest Scim.InvalidValue
-              (Just ("displayName is not compliant: " <> Text.pack err))
-
-    uref :: SAML.UserRef <- do
-      msubjid <- SAML.opaqueNameID <$$>
-        validateNameOrExtId (Scim.User.externalId user)
-
-      case msubjid of
-        Just subj -> do
-            pure $ SAML.UserRef (idp ^. SAML.idpMetadata . SAML.edIssuer) subj
-        Nothing -> throwError $
-            Scim.badRequest Scim.InvalidValue (Just "externalId is required for SAML users")
-
-    handl <- case parseHandle (Scim.User.userName user) of
-      Just x -> pure x
-      Nothing -> throwError $
-        Scim.badRequest Scim.InvalidValue (Just "userName is not compliant")
-
-    -- We check the name for validity, but only if it's present
-    mbName <- Name <$$> validateNameOrExtId (Scim.User.displayName user)
-
-    -- NB: We assume that checking that the user does _not_ exist has
-    -- already been done before -- the hscim library check does a 'get'
-    -- before a 'create'
-
+    uref :: SAML.UserRef <- case Scim.User.externalId user of
+        Just subjectTxt -> do
+            let issuer = idp ^. SAML.idpMetadata . SAML.edIssuer
+            subject <- validateSubject subjectTxt
+            pure $ SAML.UserRef issuer subject
+        Nothing -> throwError $ Scim.badRequest Scim.InvalidValue
+            (Just "externalId is required for SAML users")
+    handl <- validateHandle (Scim.User.userName user)
+    mbName <- mapM validateName (Scim.User.displayName user)
+
+    -- NB: We assume that checking that the user does _not_ exist has already been done before;
+    -- the hscim library check does a 'get' before a 'create'.
     pure $ ValidScimUser user uref handl mbName
 
+  where
+    -- Validate a subject ID (@externalId@). Right now we only support emails, but other
+    -- subject ID formats should be supported in the future.
+    validateSubject :: Text -> m SAML.NameID
+    validateSubject txt = do
+        unless (isJust (parseEmail txt)) $
```

Comment:
I think I remember now that saml2-web-sso does not care about the `NameID` format, it just takes whatever it gets.  This would give us a very simple and possibly good-enough heuristics to support both email and unspecified: if it parses as an email, use that, otherwise, use unspecified.

I think other formats are pretty obscure.

Reply:
> if it parses as an email, use that, otherwise, use unspecified.

Makes sense. I'm embarrassed that I didn't think of this.